packages/driver/src/cypress/utils.ts

```diff
@@ -399,4 +399,28 @@ export default {
   isPromiseLike (ret) {
     return ret && _.isFunction(ret.then)
   },
+
+  /* Given a chain of functions, return the actual subject. `subjectChain` might look like any of:
+   * [<input>]
+   * ['foobar', f()]
+   * [undefined, f(), f()]
+   */
+  getSubjectFromChain (subjectChain, cy) {
```

Comment:
Added some types to the PR. SubjectChain looks like so:

```
type QueryFunction = (any) => any
export type SubjectChain = [any, ...QueryFunction[]];
```

`any` in this context is a complete description - any value is valid for the first entry, including `undefined`, and QueryFunctions can take anything (one argument) and return anything (one argument).

```
Commands.addQuery (name: string, fn: () => QueryFunction)
```

Also added these types: a query is a function that returns a QueryFunction (department of redundancy department is redundant :D ).

Reply:
I think it's still useful to know that something returns itself, even if it feels a bit redundant. Eventually it would be nice to support something like

```ts
cy.get<HTMLButtonElement>('#button'). [only show relevant methods to HTMLButtonElement here]
```

Gotta start somewhere.

Reply:
Individual commands can potentially be more strictly typed - but I think you'll quickly find that most Cypress commands are so versatile that their types end up horribly messy. `.get()`, for example, can return an `HTMLElement`... or the result of an alias, which can be literally anything. '.first()`, which looks like a method for traversing HTML collections, can also accept `any[]`, or `JQuery<any>` as a previous subject. :woman_shrugging: